File: 2019/day01.exs
defmodule Fuel do
  def from_mass(mass) do
    req = div(mass, 3) - 2

    if req > 0 do
      req
    else
      0
    end
  end

  def from_mass_req(mass) do
    req = div(mass, 3) - 2

    if req > 0 do
      req + from_mass_req(req)
    else
      0
    end
  end

  def for_file(file, mass_func) do
    File.stream!(file)
    |> Enum.reduce(0, fn elem, acc ->
      fuel =
        String.trim(elem)
        |> String.to_integer()
        |> mass_func.()

      acc + fuel
    end)
  end
end

Fuel.for_file("input01.txt", &Fuel.from_mass(&1))
|> IO.puts()

Fuel.for_file("input01.txt", &Fuel.from_mass_req(&1))
|> IO.puts()

Fuel.from_mass_req(1969)
|> IO.puts()

File: 2019/day01.js
const fs = require("fs");

fileName = process.argv[2];

function fuel_for(mass) {
  return Math.max(Math.floor(mass / 3) - 2, 0);
}

function req_fuel(mass) {
  if (mass == 0) {
    return 0;
  } else {
    const fuel = fuel_for(mass);
    return fuel + req_fuel(fuel);
  }
}

data = fs
  .readFileSync(fileName, "utf-8")
  .trim()
  .split("\n")
  .reduce((acc, elem) => acc + fuel_for(+elem), 0);

// part 1
console.log(data);

data2 = fs
  .readFileSync(fileName, "utf-8")
  .trim()
  .split("\n")
  .reduce((acc, elem) => acc + req_fuel(+elem), 0);

console.log(data2);
